torch/distributed/_spmd/gm_transformation.py

```diff
@@ -0,0 +1,131 @@
+
+import torch
+import torch.nn as nn
+
+from functorch import make_fx
+from torch import fx
+from torch._inductor.compile_fx import compile_fx_inner
+from torch._inductor.decomposition import select_decomp_table
+from torch.distributed._spmd.graph_utils import OP
+from torch.distributed._spmd.iter_graph_module import IterGraphModule
+from torch.utils._pytree import tree_flatten, tree_map_only
+
+
+class InductorWrapper(nn.Module):
+    def __init__(self, gm: fx.GraphModule, enable_cudagraphs: bool) -> None:
+        super().__init__()
+        self._gm = gm
+        self._compiled: Optional[nn.Module] = None
+        self._enable_cudagraphs = enable_cudagraphs
+
+    def forward(self, *args: Any) -> Any:
+        if self._compiled is None:
+            gm = make_fx(self._gm, decomposition_table=select_decomp_table())(*args)
+            self._compiled = compile_fx_inner(
+                gm,
+                list(args),
+                cudagraphs=self._enable_cudagraphs,
+            )
+        list_args, _ = tree_flatten(args)
+        return self._compiled(list_args)
+
+
+def lower_to_inductor(
+    gm: torch.fx.GraphModule, enable_cudagraphs: bool
+) -> torch.fx.GraphModule:
+    orig_placeholders = []
+    orig_output_args: List[Any] = []
+    output = next(iter(gm.graph.nodes))
+    move_nodes = []
+
+    for node in gm.graph.nodes:
+        if node.op == OP.OUTPUT:
+            output = node
+            orig_output_args, _ = tree_flatten(node.args)
+        elif node.op == OP.PLACEHOLDER:
+            orig_placeholders.append(node)
+        else:
+            move_nodes.append(node)
+    orig_placeholders_set = set(orig_placeholders)
+
+    subgraph: torch.fx.Graph = torch.fx.Graph()
+    node_mapping: Dict[torch.fx.Node, torch.fx.Node] = {}
+    attrs = {}
+    for i, p in enumerate(orig_placeholders):
+        node_mapping[p] = subgraph.placeholder(name=f"placeholder_{i}")
+
+    for node in move_nodes:
+        if node.op == OP.GET_ATTR:
+            attrs[node.target] = getattr(gm, node.target)
+        args, kwargs = tree_map_only(
+            fx.Node, lambda n: node_mapping[n], (node.args, node.kwargs)
+        )
+        node_mapping[node] = subgraph.create_node(
+            op=node.op, target=node.target, args=args, kwargs=kwargs
+        )
+        node_mapping[node].meta = node.meta
+
+    placeholders: List[torch.fx.Node] = []
+    for placeholder in orig_placeholders:
+        new_placeholder = node_mapping[placeholder]
+        if len(new_placeholder.users) == 0:
+            # Remove unused placeholders from the subgraph. This is required as
+            # the `train_step()` has module and optimizer as the inputs which
+            # cannot be lowered to Inductor.
+            subgraph.erase_node(new_placeholder)
+        else:
+            placeholders.append(placeholder)
+    output_args = tuple(
+        node_mapping[n] for n in orig_output_args if n not in orig_placeholders_set
```

Comment:
curious, do we see cases where output args contains original inputs?

Reply:
Yes, the tracing make parameters and opt_state as input and also the updated parameters and opt_state as output. However, there can be cases where those inputs are not updated. For example, the step is not called.

Reply:
Yes, it is possible. But that is handled below instead of passing the placeholder into the subgraph and redirect it to subgraph output again.

Reply:
Updated on this. I decided to remove the placeholder check and redirect all the output even if it is directly from placeholder. The logic is cleaner.